README.md
# CSS Filter for Flutter

Use CSS filter effects on Flutter.

[![Version](https://img.shields.io/github/v/release/iofod/flutter_css_filter?label=version)](https://pub.dev/packages/css_filter)
[![Build Status](https://github.com/iofod/flutter_css_filter/workflows/build/badge.svg)](https://github.com/iofod/flutter_css_filter/actions)

[IFstruct parser](https://github.com/iofod/IFstruct-parser) supports generating flutter code from [iofod](https://www.iofod.com/) visually edited projects. We have separated its filter module code and added many preset effects to form the current CSS Filter for Flutter, which facilitates all flutter projects to use the excellent design of the Web: [CSS filter](https://developer.mozilla.org/en-US/docs/Web/CSS/filter).

## Features

- Simple and efficient
- Support multiple filter effects overlay
- Rich preset effects.
- Full platform support
- Customize your own effect
- Null-safety

## Getting started

After importing the dependent, you can use the basic filters through `CSSFilter`.

```dart
import 'package:css_filter/css_filter.dart';

CSSFilter.contrast(child: const Text('foo'), value: 1.2);
```

Support filters:

- contrast()
- grayscale()
- sepia()
- hueRotate()
- brightness()
- saturate()
- invert()
- blur()
- opacity()

You can combine and overlay these filters in flutter just as you would with a CSS filter. Use `CSSFilter.apply` in conjunction with `CSSFilterMatrix()`. Such as:

```dart
CSSFilter.apply(
  child: const Text('bar'),
  value: CSSFilterMatrix().contrast(1.2).sepia(0.8).hueRotate(90.0).invert(0.9).opacity(0.9)
);
```

**Note:** the order of the chain calls represents the order in which the filters are applied, and the order in which the filter effects are applied affects the final result.

## Use presets

The [CSSgram](https://github.com/una/CSSgram) project does an excellent job of providing the CSS world with Instagram-like filters, and [instagram.css](https://github.com/picturepan2/instagram.css) complements it with more types of filters. We've combined these two libraries to encapsulate several presets that can call through `CSSFilterPresets`.

```dart
import 'package:css_filter/css_filter.dart';

CSSFilterPresets.insXpro2(child: const Text('Your widget'));
```

Support effects:

| Effect name | Preview effect | Sample code |
| :----: | :----: | :----: |
| origin | ![origin](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/origin.jpg) | `CSSFilterPresets.origin(child: Image(...))` |
| ins1977 | ![ins1977](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/ins1977.jpg) | `CSSFilterPresets.ins1977(child: Image(...))` |
| ins1977V2 | ![ins1977V2](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/ins1977V2.jpg) | `CSSFilterPresets.ins1977V2(child: Image(...))` |
| insAden | ![insAden](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insAden.jpg) | `CSSFilterPresets.insAden(child: Image(...))` |
| insAmaro | ![insAmaro](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insAmaro.jpg) | `CSSFilterPresets.insAmaro(child: Image(...))` |
| insAshby | ![insAshby](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insAshby.jpg) | `CSSFilterPresets.insAshby(child: Image(...))` |
| insBrannan | ![insBrannan](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insBrannan.jpg) | `CSSFilterPresets.insBrannan(child: Image(...))` |
| insBrooklyn | ![insBrooklyn](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insBrooklyn.jpg) | `CSSFilterPresets.insBrooklyn(child: Image(...))` |
| insClarendon | ![insClarendon](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insClarendon.jpg) | `CSSFilterPresets.insClarendon(child: Image(...))` |
| insDogpatch | ![insDogpatch](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insDogpatch.jpg) | `CSSFilterPresets.insDogpatch(child: Image(...))` |
| insEarlybird | ![insEarlybird](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insEarlybird.jpg) | `CSSFilterPresets.insEarlybird(child: Image(...))` |
| insGingham | ![insGingham](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insGingham.jpg) | `CSSFilterPresets.insGingham(child: Image(...))` |
| insHelena | ![insHelena](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insHelena.jpg) | `CSSFilterPresets.insHelena(child: Image(...))` |
| insHudson | ![insHudson](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insHudson.jpg) | `CSSFilterPresets.insHudson(child: Image(...))` |
| insInkwell | ![insInkwell](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insInkwell.jpg) | `CSSFilterPresets.insInkwell(child: Image(...))` |
| insInkwellV2 | ![insInkwellV2](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insInkwellV2.jpg) | `CSSFilterPresets.insInkwellV2(child: Image(...))` |
| insJuno | ![insJuno](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insJuno.jpg) | `CSSFilterPresets.insJuno(child: Image(...))` |
| insKelvin | ![insKelvin](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insKelvin.jpg) | `CSSFilterPresets.insKelvin(child: Image(...))` |
| insLark | ![insLark](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insLark.jpg) | `CSSFilterPresets.insLark(child: Image(...))` |
| insLofi | ![insLofi](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insLofi.jpg) | `CSSFilterPresets.insLofi(child: Image(...))` |
| insLudwig | ![insLudwig](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insLudwig.jpg) | `CSSFilterPresets.insLudwig(child: Image(...))` |
| insMaven | ![insMaven](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insMaven.jpg) | `CSSFilterPresets.insMaven(child: Image(...))` |
| insMayfair | ![insMayfair](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insMayfair.jpg) | `CSSFilterPresets.insMayfair(child: Image(...))` |
| insMoon | ![insMoon](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insMoon.jpg) | `CSSFilterPresets.insMoon(child: Image(...))` |
| insMoonV2 | ![insMoonV2](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insMoonV2.jpg) | `CSSFilterPresets.insMoonV2(child: Image(...))` |
| insNashville | ![insNashville](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insNashville.jpg) | `CSSFilterPresets.insNashville(child: Image(...))` |
| insNashvilleV2 | ![insNashvilleV2](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insNashvilleV2.jpg) | `CSSFilterPresets.insNashvilleV2(child: Image(...))` |
| insPerpetua | ![insPerpetua](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insPerpetua.jpg) | `CSSFilterPresets.insPerpetua(child: Image(...))` |
| insPoprocket | ![insPoprocket](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insPoprocket.jpg) | `CSSFilterPresets.insPoprocket(child: Image(...))` |
| insReyes | ![insReyes](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insReyes.jpg) | `CSSFilterPresets.insReyes(child: Image(...))` |
| insRise | ![insRise](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insRise.jpg) | `CSSFilterPresets.insRise(child: Image(...))` |
| insSierra | ![insSierra](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insSierra.jpg) | `CSSFilterPresets.insSierra(child: Image(...))` |
| insSkyline | ![insSkyline](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insSkyline.jpg) | `CSSFilterPresets.insSkyline(child: Image(...))` |
| insSlumber | ![insSlumber](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insSlumber.jpg) | `CSSFilterPresets.insSlumber(child: Image(...))` |
| insStinson | ![insStinson](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insStinson.jpg) | `CSSFilterPresets.insStinson(child: Image(...))` |
| insSutro | ![insSutro](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insSutro.jpg) | `CSSFilterPresets.insSutro(child: Image(...))` |
| insToaster | ![insToaster](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insToaster.jpg) | `CSSFilterPresets.insToaster(child: Image(...))` |
| insToasterV2 | ![insToasterV2](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insToasterV2.jpg) | `CSSFilterPresets.insToasterV2(child: Image(...))` |
| insValencia | ![insValencia](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insValencia.jpg) | `CSSFilterPresets.insValencia(child: Image(...))` |
| insVesper | ![insVesper](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insVesper.jpg) | `CSSFilterPresets.insVesper(child: Image(...))` |
| insWalden | ![insWalden](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insWalden.jpg) | `CSSFilterPresets.insWalden(child: Image(...))` |
| insWaldenV2 | ![insWaldenV2](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insWaldenV2.jpg) | `CSSFilterPresets.insWaldenV2(child: Image(...))` |
| insWillow | ![insWillow](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insWillow.jpg) | `CSSFilterPresets.insWillow(child: Image(...))` |
| insXpro2 | ![insXpro2](https://raw.githubusercontent.com/iofod/flutter_css_filter/main/assets/insXpro2.jpg) | `CSSFilterPresets.insXpro2(child: Image(...))` |

Similar to `CSSFilter`, we provide the `apply` method for `CSSFilterPresets` to invoke presets consistently, and apply additionally supports setting the intensity of the filter effect used. To do so:

```dart
/// You can replace the value parameter with any of the supported effects.
CSSFilterPresets.apply(
  child: YourWidget,
  value: CSSFilterPresets.insNashville, 
  strength: 0.7
);
```

## Customize your own filter preset

`CSSFilter` provides the most basic filters, `CSSFilterPresets` collects the commonly used combinations of basic filters as presets, meaning you can also customize your own filter presets based on the basic filters.

```dart

customPreset({ required Widget child }) {
  return CSSFilter.apply(
    child: child, 
    value: CSSFilterMatrix().sepia(0.2).saturate(1.4)
  );
}

/// Used by CSSFilterPresets to set the intensity of the effect.
CSSFilterPresets.apply(
  child: YourWidget,
  value: customPreset,
  strength: 0.9
);

/// It also supports direct use.
customPreset(child: YourWidget);
```

## More examples

For more usage examples, please check out the [example project](https://github.com/iofod/flutter_css_filter/tree/main/example).


CHANGELOG.md
## 0.1.0

* initial publish.

## 0.2.0

* add example.

## 0.5.0

* add unit test.
* update docs.
* fixed isNotDefault.

## 0.9.0

* add opacity.
* update test.
* update example.
* add CI

## 1.0.0

* add CSS Instagram filters.
* update docs.
* support for custom preset.
* optimize the code structure.

## 1.0.1

* add i18n docs.
* format code.

## 1.0.2

* fix CSSFilter.blur()

## 1.0.3

* fix CSSFilterMatrix


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

// ignore: avoid_relative_lib_imports
import '../lib/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/src/base.dart
import 'dart:math' as math;
import '../css_filter.dart';
import 'utils.dart';

class FilterMatrix {
  /// Check: https://developer.mozilla.org/en-US/docs/Web/CSS/filter-function/contrast()
  static contrast({required List<double> matrix, required double value}) {
    double v = value;
    double b = (1.0 - value) * 0.5 * 255.0; // 0.5*255 => 127

    return multiplyMatrix5(matrix, <double>[
      v,
      0,
      0,
      0,
      b,
      0,
      v,
      0,
      0,
      b,
      0,
      0,
      v,
      0,
      b,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      1
    ]);
  }

  /// Formula from: https://www.w3.org/TR/filter-effects-1/#grayscaleEquivalent
  static grayscale({required List<double> matrix, required double value}) {
    double v = 1.0 - value;
    double lumR = 0.2126;
    double lumG = 0.7152;
    double lumB = 0.0722;

    return multiplyMatrix5(matrix, <double>[
      (lumR + (1 - lumR) * v),
      (lumG - lumG * v),
      (lumB - lumB * v),
      0,
      0,
      (lumR - lumR * v),
      (lumG + (1 - lumG) * v),
      (lumB - lumB * v),
      0,
      0,
      (lumR - lumR * v),
      (lumG - lumG * v),
      (lumB + (1 - lumB) * v),
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      1
    ]);
  }

  /// Formula from: https://www.w3.org/TR/filter-effects-1/#sepiaEquivalent
  static sepia({required List<double> matrix, required double value}) {
    double v = 1.0 - value;

    return multiplyMatrix5(matrix, <double>[
      (0.393 + 0.607 * v),
      (0.769 - 0.769 * v),
      (0.189 - 0.189 * v),
      0,
      0,
      (0.349 - 0.349 * v),
      (0.686 + 0.314 * v),
      (0.168 - 0.168 * v),
      0,
      0,
      (0.272 - 0.272 * v),
      (0.534 - 0.534 * v),
      (0.131 + 0.869 * v),
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      1
    ]);
  }

  /// Check: https://www.geeksforgeeks.org/css-invert-function/
  static invert({required List<double> matrix, required double value}) {
    // v * (255 - n) + (1 - v) * n => (1 - 2v) * n + 255 * v
    double v = value * 255.0;
    double k = 1.0 - 2.0 * value;

    // The fifth column n is 255.
    return multiplyMatrix5(matrix, <double>[
      k,
      0,
      0,
      0,
      v,
      0,
      k,
      0,
      0,
      v,
      0,
      0,
      k,
      0,
      v,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      1
    ]);
  }

  /// Check: https://stackoverflow.com/questions/64639589/how-to-adjust-hue-saturation-and-brightness-of-an-image-in-flutter
  static hue({required List<double> matrix, required double value}) {
    double v = math.pi * (value / 180.0);
    double cosVal = math.cos(v);
    double sinVal = math.sin(v);
    double lumR = 0.213;
    double lumG = 0.715;
    double lumB = 0.072;

    return multiplyMatrix5(matrix, <double>[
      (lumR + (cosVal * (1 - lumR))) + (sinVal * (-lumR)),
      (lumG + (cosVal * (-lumG))) + (sinVal * (-lumG)),
      (lumB + (cosVal * (-lumB))) + (sinVal * (1 - lumB)),
      0,
      0,
      (lumR + (cosVal * (-lumR))) + (sinVal * 0.143),
      (lumG + (cosVal * (1 - lumG))) + (sinVal * 0.14),
      (lumB + (cosVal * (-lumB))) + (sinVal * (-0.283)),
      0,
      0,
      (lumR + (cosVal * (-lumR))) + (sinVal * (-(1 - lumR))),
      (lumG + (cosVal * (-lumG))) + (sinVal * lumG),
      (lumB + (cosVal * (1 - lumB))) + (sinVal * lumB),
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      1
    ]);
  }

  static brightness({required List<double> matrix, required double value}) {
    double v = value;

    return multiplyMatrix5(matrix, <double>[
      v,
      0,
      0,
      0,
      0,
      0,
      v,
      0,
      0,
      0,
      0,
      0,
      v,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      1
    ]);
  }

  /// Check: https://docs.rainmeter.net/tips/colormatrix-guide/
  static saturate({required List<double> matrix, required double value}) {
    return FilterMatrix.grayscale(matrix: matrix, value: 1.0 - value);
  }

  static opacity({required List<double> matrix, required double value}) {
    return multiplyMatrix5(matrix, <double>[
      1,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      value,
      0,
      0,
      0,
      0,
      0,
      1
    ]);
  }
}

// Allows matrix multiplication.
final filterTypeMap = {
  'contrast': FilterMatrix.contrast,
  'grayscale': FilterMatrix.grayscale,
  'hueRotate': FilterMatrix.hue,
  'brightness': FilterMatrix.brightness,
  'saturate': FilterMatrix.saturate,
  'opacity': FilterMatrix.opacity,
};

// Not superimposed on the original matrix.
final filterAloneMap = {
  'sepia': CSSFilter.sepia,
  'invert': CSSFilter.invert,
  'blur': CSSFilter.blur
};


lib/src/util.dart
import 'package:flutter/material.dart';

List<double> baseMatrix() {
  return <double>[
    1,0,0,0,0,
    0,1,0,0,0,
    0,0,1,0,0,
    0,0,0,1,0,
    0,0,0,0,1
  ];
}

/// Check: https://github.com/openkraken/kraken/blob/main/kraken/lib/src/css/filter.dart
/// Calc 5x5 matrix multiplication.
List<double> multiplyMatrix5(List<double> a, List<double> b) {
  if (a.length != b.length) {
    throw FlutterError('Matrix length should be same.');
  }

  if (a.length != 25) {
    throw FlutterError('Matrix5 size is not correct.');
  }

  var a00 = a[0];
  var a01 = a[1];
  var a02 = a[2];
  var a03 = a[3];
  var a04 = a[4];
  var a10 = a[5];
  var a11 = a[6];
  var a12 = a[7];
  var a13 = a[8];
  var a14 = a[9];
  var a20 = a[10];
  var a21 = a[11];
  var a22 = a[12];
  var a23 = a[13];
  var a24 = a[14];
  var a30 = a[15];
  var a31 = a[16];
  var a32 = a[17];
  var a33 = a[18];
  var a34 = a[19];
  var a40 = a[20];
  var a41 = a[21];
  var a42 = a[22];
  var a43 = a[23];
  var a44 = a[24];

  var b00 = b[0];
  var b01 = b[1];
  var b02 = b[2];
  var b03 = b[3];
  var b04 = b[4];
  var b10 = b[5];
  var b11 = b[6];
  var b12 = b[7];
  var b13 = b[8];
  var b14 = b[9];
  var b20 = b[10];
  var b21 = b[11];
  var b22 = b[12];
  var b23 = b[13];
  var b24 = b[14];
  var b30 = b[15];
  var b31 = b[16];
  var b32 = b[17];
  var b33 = b[18];
  var b34 = b[19];
  var b40 = b[20];
  var b41 = b[21];
  var b42 = b[22];
  var b43 = b[23];
  var b44 = b[24];

  return [
    a00 * b00 + a01 * b10 + a02 * b20 + a03 * b30 + a04 * b40,
    a00 * b01 + a01 * b11 + a02 * b21 + a03 * b31 + a04 * b41,
    a00 * b02 + a01 * b12 + a02 * b22 + a03 * b32 + a04 * b42,
    a00 * b03 + a01 * b13 + a02 * b23 + a03 * b33 + a04 * b43,
    a00 * b04 + a01 * b14 + a02 * b24 + a03 * b34 + a04 * b44,
    a10 * b00 + a11 * b10 + a12 * b20 + a13 * b30 + a14 * b40,
    a10 * b01 + a11 * b11 + a12 * b21 + a13 * b31 + a14 * b41,
    a10 * b02 + a11 * b12 + a12 * b22 + a13 * b32 + a14 * b42,
    a10 * b03 + a11 * b13 + a12 * b23 + a13 * b33 + a14 * b43,
    a10 * b04 + a11 * b14 + a12 * b24 + a13 * b34 + a14 * b44,
    a20 * b00 + a21 * b10 + a22 * b20 + a23 * b30 + a24 * b40,
    a20 * b01 + a21 * b11 + a22 * b21 + a23 * b31 + a24 * b41,
    a20 * b02 + a21 * b12 + a22 * b22 + a23 * b32 + a24 * b42,
    a20 * b03 + a21 * b13 + a22 * b23 + a23 * b33 + a24 * b43,
    a20 * b04 + a21 * b14 + a22 * b24 + a23 * b34 + a24 * b44,
    a30 * b00 + a31 * b10 + a32 * b20 + a33 * b30 + a34 * b40,
    a30 * b01 + a31 * b11 + a32 * b21 + a33 * b31 + a34 * b41,
    a30 * b02 + a31 * b12 + a32 * b22 + a33 * b32 + a34 * b42,
    a30 * b03 + a31 * b13 + a32 * b23 + a33 * b33 + a34 * b43,
    a30 * b04 + a31 * b14 + a32 * b24 + a33 * b34 + a34 * b44,
    a40 * b00 + a41 * b10 + a42 * b20 + a43 * b30 + a44 * b40,
    a40 * b01 + a41 * b11 + a42 * b21 + a43 * b31 + a44 * b41,
    a40 * b02 + a41 * b12 + a42 * b22 + a43 * b32 + a44 * b42,
    a40 * b03 + a41 * b13 + a42 * b23 + a43 * b33 + a44 * b43,
    a40 * b04 + a41 * b14 + a42 * b24 + a43 * b34 + a44 * b44,
  ];
}

ColorFilter toColorFilterMatrix(matrix) {
  return ColorFilter.matrix(matrix.sublist(0, 20));
}

Widget execFilterSample(matrix, child) {
  return ColorFiltered(colorFilter: toColorFilterMatrix(matrix), child: child);
}

bool isNotNegative(double v) {
  return v > 0.0;
}

bool isNotDefault(double v) {
  return v != 1.0 && v >= 0.0;
}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:css_filter/css_filter.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'CSS Filter Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

double width = 360.0;

Widget pic(name) {
  int i = 0;

  return Container(
      padding:
          const EdgeInsets.only(top: 10.0, right: 0.0, bottom: 10.0, left: 0.0),
      child: Stack(
        children: [
          Positioned(
              child: Image(
            image: const NetworkImage(
                'https://static.iofod.com/common/buildings/1621839293951_castle_1280x852.jpg'),
            width: width / 2.5,
          )),
          ...(name ?? '').split(',').map((str) {
            double top = i * 30.0;

            i++;

            return Positioned(
                top: top,
                child: Container(
                    decoration: const BoxDecoration(
                        color: Color.fromARGB(151, 61, 61, 61)),
                    padding: const EdgeInsets.all(6.0),
                    child: Text(str,
                        style: const TextStyle(
                            fontSize: 12.0,
                            color: Color.fromARGB(255, 251, 255, 9)))));
          }).toList(),
        ],
      ));
}

Widget customPreset({required Widget child}) {
  return CSSFilter.apply(
      child: child,
      value: CSSFilterMatrix().sepia(0.5).hueRotate(-30.0).saturate(1.4));
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    width = MediaQuery.of(context).size.width;

    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SingleChildScrollView(
          scrollDirection: Axis.vertical,
          child: CSSFilter.blur(
              value: 0.0,
              child: Container(
                  padding: const EdgeInsets.all(16.0),
                  width: MediaQuery.of(context).size.width,
                  child: Wrap(
                    alignment: WrapAlignment.spaceEvenly,
                    runAlignment: WrapAlignment.spaceEvenly,
                    children: [
                      pic('Original image'),

                      CSSFilter.contrast(
                          child: pic('contrast(1.3)'), value: 1.3),
                      CSSFilter.grayscale(
                          child: pic('grayscale(0.5)'), value: 0.5),
                      CSSFilter.sepia(child: pic('sepia(0.5)'), value: 0.5),
                      CSSFilter.hueRotate(
                          child: pic('hueRotate(60.0)'), value: 60.0),
                      CSSFilter.brightness(
                          child: pic('brightness(1.3)'), value: 1.3),
                      CSSFilter.saturate(
                          child: pic('saturate(2.0)'), value: 2.0),
                      CSSFilter.invert(child: pic('invert(0.9)'), value: 0.9),

                      ClipRect(
                        child:
                            CSSFilter.blur(child: pic('blur(1.5)'), value: 1.5),
                      ),

                      CSSFilter.opacity(child: pic('opacity(0.5)'), value: 0.5),

                      CSSFilter.apply(
                          child: pic('contrast(1.5),sepia(0.2)'),
                          value: CSSFilterMatrix().contrast(1.3).sepia(0.2)),
                      CSSFilter.apply(
                          child: pic('hueRotate(60.0),grayscale(0.5)'),
                          value:
                              CSSFilterMatrix().hueRotate(60.0).grayscale(0.5)),
                      // CSSFilter.apply(
                      //     child: pic('brightness(1.2),saturate(0.8),hueRotate(20.0)'),
                      //     value: CSSFilterMatrix().brightness(1.2).saturate(0.8).hueRotate(20.0)),
                      CSSFilter.apply(
                          child: pic(
                              'contrast(1.2),sepia(0.8),hueRotate(90.0),invert(0.9),opacity(0.9)'),
                          value: CSSFilterMatrix()
                              .contrast(1.2)
                              .sepia(0.8)
                              .hueRotate(90.0)
                              .invert(0.9)
                              .opacity(0.9)),

                      CSSFilterPresets.ins1977(child: pic('ins1977')),
                      CSSFilterPresets.ins1977V2(child: pic('ins1977V2')),
                      CSSFilterPresets.apply(
                          child: pic('ins1977V2 customPreset'),
                          value: customPreset,
                          strength: 0.9),
                      CSSFilterPresets.insAden(child: pic('insAden')),
                      CSSFilterPresets.insAmaro(child: pic('insAmaro')),
                      CSSFilterPresets.insAshby(child: pic('insAshby')),
                      CSSFilterPresets.insBrannan(child: pic('insBrannan')),
                      CSSFilterPresets.insBrooklyn(child: pic('insBrooklyn')),
                      CSSFilterPresets.insClarendon(child: pic('insClarendon')),
                      CSSFilterPresets.insDogpatch(child: pic('insDogpatch')),
                      CSSFilterPresets.insEarlybird(child: pic('insEarlybird')),
                      CSSFilterPresets.insGingham(child: pic('insGingham')),
                      CSSFilterPresets.insHelena(child: pic('insHelena')),
                      CSSFilterPresets.insHudson(child: pic('insHudson')),
                      CSSFilterPresets.insInkwell(child: pic('insInkwell')),
                      CSSFilterPresets.insInkwellV2(child: pic('insInkwell')),
                      CSSFilterPresets.insJuno(child: pic('insJuno')),
                      CSSFilterPresets.insKelvin(child: pic('insKelvin')),
                      CSSFilterPresets.insLark(child: pic('insLark')),
                      CSSFilterPresets.insLofi(child: pic('insLofi')),
                      CSSFilterPresets.insLudwig(child: pic('insLudwig')),
                      CSSFilterPresets.insMaven(child: pic('insMaven')),
                      CSSFilterPresets.insMayfair(child: pic('insMayfair')),
                      CSSFilterPresets.insMoon(child: pic('insMoon')),
                      CSSFilterPresets.insMoonV2(child: pic('insMoonV2')),
                      CSSFilterPresets.insNashville(child: pic('insNashville')),
                      CSSFilterPresets.insNashvilleV2(
                          child: pic('insNashvilleV2')),
                      CSSFilterPresets.insPerpetua(child: pic('insPerpetua')),
                      CSSFilterPresets.insPoprocket(child: pic('insPoprocket')),
                      CSSFilterPresets.insReyes(child: pic('insReyes')),
                      CSSFilterPresets.insRise(child: pic('insRise')),
                      CSSFilterPresets.insSierra(child: pic('insSierra')),
                      CSSFilterPresets.insSkyline(child: pic('insSkyline')),
                      CSSFilterPresets.insSlumber(child: pic('insSlumber')),
                      CSSFilterPresets.insStinson(child: pic('insStinson')),
                      CSSFilterPresets.insSutro(child: pic('insSutro')),
                      CSSFilterPresets.insToaster(child: pic('insToaster')),
                      CSSFilterPresets.insToasterV2(child: pic('insToasterV2')),
                      CSSFilterPresets.insValencia(child: pic('insValencia')),
                      CSSFilterPresets.insVesper(child: pic('insVesper')),
                      CSSFilterPresets.insWalden(child: pic('insWalden')),
                      CSSFilterPresets.insWaldenV2(child: pic('insWaldenV2')),
                      CSSFilterPresets.insWillow(child: pic('insWillow')),
                      CSSFilterPresets.insXpro2(child: pic('insXpro2')),

                      CSSFilterPresets.apply(
                          child: pic('apply insToaster strength'),
                          value: CSSFilterPresets.insToaster,
                          strength: 0.7),

                      CSSFilterPresets.apply(
                          child: Opacity(
                              opacity: 0.7,
                              child: pic('apply ins1977 alphaBlending')),
                          value: CSSFilterPresets.ins1977,
                          strength: 0.8,
                          alphaBlending: 0.7),

                      pic('Original image')
                    ],
                  )))),
    );
  }
}
